File: 496_next_greater_element_i/next_greater_element_i.py
from typing import *


class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        stack = []
        table = {}
        for i in range(len(nums2)):
            while stack and nums2[i] > nums2[stack[-1]]:
                table[nums2[stack.pop()]] = nums2[i]
            stack.append(i)
        ans = []
        for i in nums1:
            ans.append(table.get(i, -1))
        return ans

File: 496_next_greater_element_i/next_greater_element_i.java
import java.util.Stack;
import java.util.HashMap;

class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        Stack<Integer> stack = new Stack<Integer>();
        HashMap<Integer, Integer> map = new HashMap<Integer, Integer>();
        for (int i = 0; i < nums2.length; i++) {
            while (!stack.empty() && nums2[i] > nums2[stack.peek()])
                map.put(nums2[stack.pop()], nums2[i]);
            stack.push(i);
        }
        int[] ans = new int[nums1.length];
        for (int i = 0; i< nums1.length; i++) {
            ans[i] = map.containsKey(nums1[i]) ? map.get(nums1[i]) : -1;
        }
        return ans;
    }
}